source/common/config/delta_subscription_state.cc

```diff
@@ -53,6 +62,28 @@ void DeltaSubscriptionState::updateSubscriptionInterest(
     names_added_.erase(r);
     names_removed_.insert(r);
   }
+  // If we unsubscribe from wildcard resource, drop all the resources that came from wildcard from
+  // cache.
+  if (cur_removed.contains(Wildcard)) {
+    wildcard_resource_state_.clear();
+  }
+  // Check if this is a legacy wildcard subscription request. If we repeatedly call this function
+  // with empty cur_added and cur_removed, we keep the legacy wildcard subscription.
+  if (is_legacy_wildcard_) {
```

Comment:
@adisuissa: I remember you added an explicit wildcard mode in delta state; was there are a reason you decided against determining it implicitly, like below?

Reply:
If I were to guess, then I would say that before my changes the wildcard mode was an immutable trait of delta state, and we knew at a time of delta state creation whether it's going to be in wildcard mode or not, so why not leverage that knowledge? That certainly led to a simpler code and less work being done when updating a subscription interest. And now the wildcard mode is not an immutable trait, so determining the mode at a creation time does not buy us much any more.

Reply:
IIRC we needed that when a client reconnects on a stream. In wildcard mode the client sends a wildcard/empty-list, and in non-wildcard mode it sends all the resource names.
Looking [here](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#incremental-xds) it was needed to support this case:
```
Note that for “wildcard” subscriptions (see How the client specifies what resources to return for details),
the request must either specify “*” in the resource_names_subscribe field or (legacy behavior)
the request must have no resources in both resource_names_subscribe and resource_names_unsubscribe.
```